### DISCUSSION TITLE
How to print the request body ?

### DISCUSSION THREAD

**Discussion by paulzhang5511** [rwf2/Rocket#1905] [Questions] [answered]
```rust
        .attach(AdHoc::on_request("PUT Rewriter", |req, _| {
            Box::pin(async move {
                println!("    => Incoming request: {}", req);
            })
        }))
        .attach(AdHoc::on_response("Response Rewriter", |req, res| {
            Box::pin(async move {
                if req.uri().path() == "/" {
                    println!("    => Rewriting response body.");
                    let s = res.body_mut().to_string().await.unwrap();
                    println!("body --> {}", s);
                    res.set_sized_body(None, Cursor::new(s));
                }
            })
        }))

```

The above code can only print the response body, I don't know how to print the request body.

**Comment by jebrosen** [comment, marked as answer]
The second parameter to an `on_request` fairing (currently `_` in your posted example) is an `&mut Data`, which can be used to [peek](https://api.rocket.rs/v0.5-rc/rocket/data/struct.Data.html#method.peek) at the beginning of the request data.

If you do need to read the entire data, you should use a [data guard](https://rocket.rs/v0.5-rc/guide/requests/#body-data) on the appropriate route instead - but see also #775 regarding some of the limitations. In particular, it's not easy out-of-the-box to both print the body data and do something else with it, such as parsing JSON input.
